cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -1,39 +1,74 @@
 use std::path::PathBuf;
 
+use chrono::{DateTime, Utc};
+use flox_core::canonical_path::CanonicalPath;
 use thiserror::Error;
 
 use super::build::ManifestBuilder;
 use super::catalog::Client;
 use super::git::GitCommandProvider;
+// use crate::flox::test_helpers::flox_instance;
+use crate::flox::Flox;
 use crate::flox::FloxhubToken;
 use crate::models::environment::managed_environment::ManagedEnvironment;
 use crate::models::environment::path_environment::PathEnvironment;
+use crate::models::environment::Environment;
+// use crate::models::lockfile;
 use crate::models::lockfile::Lockfile;
+// use crate::providers::build;
+// use crate::models::lockfile::Lockfile;
+use crate::providers::git::GitProvider;
 
 pub enum PublishEnvironment {
     Path(PathEnvironment),
     Managed(ManagedEnvironment),
 }
 
+#[derive(Debug, Error)]
+pub enum PublishError {
+    #[error("This type of environment is not supported for publishing")]
+    UnsupportedEnvironment,
+    #[error("The environment must be locked to publish")]
+    UnlockedEnvironment,
+
+    #[error("The environment is in an unsupported state for publishing")]
+    UnsupportEnvironmentState { msg: String },
+
+    #[error("Could not identify user from authentication info")]
+    Unauthenticated,
+}
+
 /// The `Publish` trait describes the high level behavior of publishing a package to a catalog.
 /// Authentication, upload, builds etc, are implementation details of the specific provider.
 /// Modeling the behavior as a trait allows us to swap out the provider, e.g. a mock for testing.
 pub trait Publish {
     fn publish(
         &self,
+        flox: &Flox,
         client: Client,
-        builder: impl ManifestBuilder,
         environment: PublishEnvironment,
-        package: String,
-    ) -> Result<(), String>;
+        builder: impl ManifestBuilder,
+        package: &str,
+    ) -> Result<(), PublishError>;
+}
+
+/// Simple struct to hold the information of a locked URL.
+#[derive(Debug)]
```

Comment:
suggestion (nb): more derives

```suggestion
#[derive(Debug, Clone, PartialEq, Eq)]
```